### DISCUSSION TITLE
Does npm install dependency updates under any circumstance when a valid package-lock.json exists?

### DISCUSSION THREAD

**Discussion by kcmr** [npm/feedback#990] [CLI] [answered]
I'm considering using `npm install` instead of `npm ci` in our CI due to its faster execution when we have `node_modules` (restored from cache), as we have a limitation in build minutes.

I understand that `npm ci` is designed for this specific purpose, as it ensures a deterministic dependency tree based on the lock files. However, I believe the behavior of `npm install` has changed since version 7 and `lockfileVersion >= 2`. Unfortunately, I'm having trouble finding concise documentation or references about this, apart from this [quote from the GitHub blog](https://github.blog/2020-10-13-presenting-v7-0-0-of-the-npm-cli/):

> [package-lock v2 and support for yarn.lock](https://blog.npmjs.org/post/621733939456933888/npm-v7-series-why-keep-package-lockjson): Our new package-lock format will unlock the ability to do deterministically reproducible builds and includes everything npm will need to fully build the package tree.

In the past, when you used `npm install`, you could end up installing unwanted updates for dependencies. I think this is no longer the case. Can you confirm if my understanding is right?

Considering the benefits of using `npm ci` over `npm install`, such as the validation of the package-lock.json file, are there any other reasons to prefer `npm ci` over `npm install` in CI environments when it comes to handling dependencies?

Thank you.

**Comment by ljharb** [comment]
I would expect `npm install` to always potentially do updates, and only `npm ci` to ensure the lockfile is strictly obeyed.

**Reply by kcmr** [reply]
Yes, that's what I always had expected, but I think it no longer updates dependencies.

I've performed the following test in my machine:
- delete the npm cache (`npm cache clear --force`)
- remove `node_modules` from a project
- run `npm install`

All the exact versions were preserved, and no changes were introduced in the `package-lock.json` file.
My `lockfileVersion` is 3

**Reply by wraithgar** [reply, marked as answer]
A bare `npm install` with no options won't update your package-lock if it doesn't have to.  This is a byproduct of the way npm favors keeping what's in your tree if it can and not part of the `npm install` contract.  If you `npm install <spec>` then it will definitely change things.  Older lockfile versions will also potentially result in a change to your tree I believe.  Optional dependencies may or may not be added or removed.  Packages may have been deprecated since your lockfile since it was created.

Most importantly, you have no way of knowing during `npm install` if your package.json and package-lock.json files agree.  `npm install` assumes it can fix the `package-lock.json` and will do so if it needs to, with no way for you to know beforehand (outside of `npm ci`) if this will happen.

Consider this

```sh
$ npm init -y
$ npm install abbrev
$ npm pkg set dependencies.semver='^7.0.0'
```

At this point `npm ci` will fail, while `npm install` will succeed and add semver to your package lock.

So this is a case of "past performance does not guarantee future results".  The current `npm install` behavior doesn't change your tree in *many* cases but it is not something I would rely on.

TLDR `npm install` reserves the right to change your `package.json` and `package-lock.json`.
